CMSgov/easi-app
# EASI-2234 - Send a single notification email, regardless of the number of recipients

# EASI-2234

## Changes and Description

- The only change in _functionality_ is that all notification recipients are listed as addressees on a single email, instead of separate emails being sent to each recipient.
- Under the hood, this required changing the `sender` interface to take a slice for `toAddresses`, instead of a single address. I also went ahead and made `ccAddresses` a slice as well. Most of this PR is just adjusting code to account for this change.

## How to test this change

* Set up your environment to test with multiple recipients locally
    * Connect to the VPN (in order to be able to connect to CEDAR LDAP).
    * In `pkg/server/routes.go`, comment out the block that sets `cedarLDAPClient = local.NewCedarLdapClient(s.logger)`.
    * In `pkg/services/config.go`, change `checkBoolFeatureFlag()` to unconditionally return true.
    * In `src/views/FlagsWrapper/index.tsx`, set the default value for `notifyMultipleRecipients` to `true` instead of `false`.
    * In `docker-compose.yml`, change the values defined for the `GRT_EMAIL` and `IT_INVESTMENT_EMAIL` environment variables to different values so they can be distinguished.
    * Start up the application locally
* Log in to the application with a _valid_ EUA ID (invalid EUA IDs, such as ABCD, will cause issues due to not being able to look up their email addresses).
* Create a system intake. 
* Through the admin view, go to Actions, then "Issue a lifecycle ID".
* Select a mix of notification recipients (some selected, some not), then click "Complete action and send email".
* Open Mailcatcher at `localhost:1080` - you should see a single notification email, with the addressees being those selected when issuing the LCID.

![test email - UI](https://user-images.githubusercontent.com/92891039/187771342-ddd297cb-4ce7-49d0-949e-dfc9466429ec.PNG)

![test email - recipients](https://user-images.githubusercontent.com/92891039/187771355-63ee4fbb-6107-46bb-b327-96e0463d6bc4.PNG)

## Notes
Even if some of the recipients' addresses are invalid and the email bounces, the call to the AWS SES SDK will _not_ return an error. See [EASI-2340](https://jiraent.cms.gov/browse/EASI-2340).
